examples/htool_jtag.c

```diff
@@ -0,0 +1,199 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "htool_jtag.h"
+
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "host_commands.h"
+#include "htool.h"
+#include "htool_cmd.h"
+
+char JTAG_TEST_BYPASS_PATTERN_DEFAULT_VALUE[] =
+    // PRBS9 with '0' bit added at the beginning to make it exactly 64 bytes
+    "0x42 0x30 0x9c 0xab 0xd 0xe9 0xb9 0x14 0x2b 0x4f 0xd9 0x25 0xbf 0x26 0xa6 "
+    "0x60 0x31 0x94 0x69 0x7f 0x45 0x8e 0xb2 0xcf 0x1f 0x74 0x1a 0xdb 0xb0 "
+    "0x5a 0xfa 0xa8 0x14 0xaf 0x2e 0xe0 0x73 0xa4 0xf5 0xd4 0x48 0x67 0xb 0xdb "
+    "0x34 0x3b 0xc3 0xfe 0xf 0x7c 0x5c 0xc8 0x25 0x3b 0x47 0x9f 0x36 0x2a 0x47 "
+    "0x1b 0x57 0x13 0x11 0x0";
+
+static int jtag_read_idcode(struct libhoth_device *dev,
+                            const struct htool_invocation *inv) {
+  uint32_t clk_idiv;
+
+  if (htool_get_param_u32(inv, "clk_idiv", &clk_idiv)) {
+    return -1;
+  }
+
+  if (clk_idiv > UINT16_MAX) {
+    fprintf(stderr, "Clock divisor value too large. Expected <= %u\n",
+            UINT16_MAX);
+    return -1;
+  }
+
+  const struct ec_request_jtag_operation request = {
+      .clk_idiv = (uint16_t)clk_idiv,
+      .operation = EC_JTAG_OP_READ_IDCODE,
+  };
+
+  struct ec_response_jtag_read_idcode_operation response;
+  size_t response_length = 0;
+  int ret = htool_exec_hostcmd(
+      dev, EC_CMD_BOARD_SPECIFIC_BASE + EC_PRV_CMD_HOTH_JTAG_OPERATION,
+      /*version=*/0, &request, sizeof(request), &response, sizeof(response),
+      &response_length);
+
+  if (ret != 0) {
+    fprintf(stderr, "HOTH_JTAG_OPERATION error code: %d\n", ret);
+    return -1;
+  }
+
+  if (response_length != sizeof(response)) {
+    fprintf(stderr,
+            "HOTH_JTAG_OPERATION expected exactly %zu reseponse bytes, got %zu",
+            sizeof(response), response_length);
+    return -1;
+  }
+
+  printf("IDCODE: 0x%08x\n", response.idcode);
+  return 0;
+}
+
+// Parse exactly `expected_bytes_count` byte tokens from `param_value` string
+// into `byte_sequence`.
+// If fewer than `expected_bytes_count` byte tokens are present in
+// `param_value`, return code indicates error.
+// If more than `expected_bytes_count` byte tokens are present in `param_value`,
+// extra bytes are ignored and no error is indicated
```

Comment:
I am assuming this behavior is okay. If an error needs to be returned, I can change the code here.

Reply:
Someone trying to send a pattern greater than 64 bytes might be confused why htool is truncating their input, in which case a printed warning would clarify matters. Likewise, someone trying to send exactly 64 bytes should be made aware when their input is unintentionally longer.

Reply:
Fixed. The new changes should have the function accept exactly `expected_bytes_count` bytes in form of string tokens (no more and no less)